src/middleware/compression.rs

```diff
@@ -0,0 +1,281 @@
+            self.encode(res, encoding)
+        }
+    }
+}
+
+/// Middleware for handling incoming request decompression.
+#[derive(Debug, Default)]
+pub struct Decompression {}
+
+impl Decompression {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn parse_encoding(&self, s: &str) -> Result<Encoding, Error> {
+        match s {
+            "gzip" => Ok(Encoding::Gzip),
+            "deflate" => Ok(Encoding::Deflate),
+            "identity" => Ok(Encoding::Identity),
+            "br" => Ok(Encoding::Deflate),
+            _ => Err(Error::from(StatusCode::UNSUPPORTED_MEDIA_TYPE)),
+        }
+    }
+
+    fn decode(&self, req: &mut Request) -> Result<(), Error> {
+        // TODO: test and make sure this works
+        let encodings = if let Some(hval) = req.headers().get_all(CONTENT_ENCODING).iter().next() {
+            let hval = hval.to_str().unwrap();
+            let mut encodings: Vec<Encoding> = Vec::new();
+            for v in hval.split(',').map(str::trim) {
+                if let Some(cap) = RE_CONTENT_ENCODING.captures(v) {
+                    encodings.push(self.parse_encoding(cap.name("a").unwrap().as_str())?);
+                } else {
+                    return Err(Error::from(StatusCode::BAD_REQUEST));
+                }
+            }
+            encodings.reverse(); // apply decodings in reverse order of which they were applied
+            encodings
+        } else {
+            return Ok(());
+        };
+
+        for encoding in encodings {
+            match encoding {
+                Encoding::Gzip => {
+                    let mut body = Body::empty();
+                    std::mem::swap(&mut body, req.body_mut());
+                    let stream = gzip::DecompressedGzipStream::new(body);
+                    *req.body_mut() = Body::from_stream(stream);
+                }
+                Encoding::Deflate => {
+                    let mut body = Body::empty();
+                    std::mem::swap(&mut body, req.body_mut());
+                    let stream = zlib::DecompressedZlibStream::new(body);
+                    *req.body_mut() = Body::from_stream(stream);
+                }
+                Encoding::Identity => (),
+                Encoding::Br => {
+                    let mut body = Body::empty();
+                    std::mem::swap(&mut body, req.body_mut());
+                    let stream = brotli::DecompressedBrotliStream::new(body);
+                    *req.body_mut() = Body::from_stream(stream);
+                }
+            }
+        }
+
+        // strip the content-encoding header
+        req.headers_mut().remove(CONTENT_ENCODING).unwrap();
+
+        Ok(())
+    }
+}
+
+impl<Data: Send + Sync + 'static> Middleware<Data> for Decompression {
+    fn handle<'a>(&'a self, cx: Context<Data>, next: Next<'a, Data>) -> BoxFuture<'a, Response> {
+        box_async! {
+//            match self.decode(cx.request()) { // <- There needs to be a way to grab a mut reference to the request
```

Comment:
The logic is pretty much done for the decompression middleware, but this is a blocker.

Currently, there's no existing way to get a `&mut Request` handle from the Request context, and while there is a way to take the body through `take_body(&mut self)`, there's no way to set it through some existing method like `set_body(&mut self)` or so.

https://docs.rs/tide/0.2.0/tide/struct.Context.html#methods